File: src/patterns/behavioral/mediator/mediator.go
package main

import "fmt"

// Mediator
type IMediator interface {
	Notify(sender interface{}, event string) string
}

type BookingMediator struct {
	bookingComponent *BookingComponent
	notificationComponent *NotificationComponent
}

func NewBookingMediator(bookingComponent *BookingComponent, notificationComponent *NotificationComponent) *BookingMediator {
	mediator := &BookingMediator{bookingComponent, notificationComponent}
	bookingComponent.SetMediator(mediator)
	notificationComponent.SetMediator(mediator)
	return mediator
}

func (m *BookingMediator) Notify(sender interface{}, event string) string {
	if event == "ReserveRoom" {
		return m.notificationComponent.SendMessage(m.bookingComponent.username, m.bookingComponent.reservationNumber)
	}

	return "Unknown notification"
}

// Booking Component
type BookingComponent struct {
	reservationNumber int
	username string
	mediator *BookingMediator
}

func (c *BookingComponent) SetMediator(mediator *BookingMediator) {
	c.mediator = mediator
}

func (c *BookingComponent) ReserveRoom() string {
	return c.mediator.Notify(c, "ReserveRoom")
}

// Notification Component
type NotificationComponent struct {
	mediator *BookingMediator
}

func (c *NotificationComponent) SetMediator(mediator *BookingMediator) {
	c.mediator = mediator
}

func (c *NotificationComponent) SendMessage(username string, reservationNumber int) string {
	return fmt.Sprintf("The message is sent to %v, the reservation is confirmed and the number of the reservation is %v", username, reservationNumber)
}

func main(){
	notifier := &NotificationComponent{}
	booking := &BookingComponent{username: "nikita", reservationNumber: 123}

	NewBookingMediator(booking, notifier)
	fmt.Println(booking.ReserveRoom())
}

File: src/patterns/behavioral/mediator/mediator.ts
// Without the mediator

// // Booking class
// class BookingComponent {
//     notifier: NotificationComponent
//     username: string
//     reservationNumber: number
//     constructor(notifier: NotificationComponent, username: string, reservationNumber: number) {
//         this.notifier = notifier
//         this.username = username
//         this.reservationNumber = reservationNumber
//     }

//     setUsername(username:string){
//         this.username = username
//     }

//     setReservationNumber(reservationNumber:number){
//         this.reservationNumber = reservationNumber
//     }

//     reserveNumber(){
//         this.notifier.sendNotification(this.username, this.reservationNumber)
//     }
// }

// // Notification component
// class NotificationComponent {
//   sendNotification(username: string, reservationNumber: number): string {
//     return `Reservation acquired for user ${username} with the number of ${reservationNumber}`;
//   }
// }

// With the mediator

// Mediator
interface IMediator {
  notify(sender: object, event: string): string;
}

class BookingMediator implements IMediator {
  bookingComponent: BookingComponent;
  notifierComponent: NotificationComponent;

  constructor(
    bookingComponent: BookingComponent,
    notifierComponent: NotificationComponent
  ) {
    this.bookingComponent = bookingComponent;
    this.bookingComponent.setMediator(this);
    this.notifierComponent = notifierComponent;
    this.notifierComponent.setMediator(this);
  }

  notify(sender: object, event: string): string {
    if (event === "ReserveNumber") {
      return this.notifierComponent.sendNotification(
        this.bookingComponent.username,
        this.bookingComponent.reservationNumber
      );
    }

    return "Unknown event";
  }
}

// Booking class
class BookingComponent {
  username: string;
  reservationNumber: number;
  mediator: IMediator | null = null;
  constructor(username: string, reservationNumber: number) {
    this.username = username;
    this.reservationNumber = reservationNumber;
  }

  setMediator(mediator: IMediator) {
    this.mediator = mediator;
  }

  setUsername(username: string) {
    this.username = username;
  }

  setReservationNumber(reservationNumber: number) {
    this.reservationNumber = reservationNumber;
  }

  reserveNumber(): string {
    return this.mediator!.notify(this, "ReserveNumber");
  }
}

// Notification component
class NotificationComponent {
  mediator: IMediator | null = null;

  setMediator(mediator: IMediator) {
    this.mediator = mediator;
  }

  sendNotification(username: string, reservationNumber: number): string {
    return `Reservation acquired for user ${username} with the number of ${reservationNumber}`;
  }
}

const booking = new BookingComponent("nikita", 123);
const notifier = new NotificationComponent();
const mediator = new BookingMediator(booking, notifier);

console.log(booking.reserveNumber());
